File: ShadowFrameworkLite20/src/shadow/operational/SFPrimitiveArrayValueList.java
package shadow.operational;

import shadow.math.SFValuenf;
import shadow.pipeline.SFPrimitiveArray;
import shadow.pipeline.expression.SFExpressionValuesList;
import shadow.system.SFArray;

public class SFPrimitiveArrayValueList implements SFExpressionValuesList{

	private int[] primitiveIndices;
	private SFPrimitiveArray array;
	private int position=0;
	private SFArray<SFValuenf> arrayData;
	
	public SFPrimitiveArrayValueList(int[] primitiveIndices, SFPrimitiveArray array) {
		super();
		this.primitiveIndices = primitiveIndices;
		this.array = array;
	}
	
	public void setGridIndex(int gridIndex){
		position=array.getPrimitive().getIndicesPositions()[gridIndex];
		arrayData=array.getPrimitiveData(gridIndex);
	}
	
	@Override
	public SFValuenf generateValue() {
		return arrayData.generateSample();
	}
	
	public int getDataIndex(int index){
		return primitiveIndices[position+index];
	}
	
	@Override
	public SFValuenf getValue(int index) {
		SFValuenf value=arrayData.generateSample();
		arrayData.getElement(getDataIndex(index), value);
		return value;
	}
	
}

File: trunk/ShadowFrameworkJS/src/shadow/operational/SFPrimitiveArrayValueList.js


function SFPrimitiveArrayValueList(primitiveIndices,array){
	this.primitiveIndices = primitiveIndices;
	this.array = array;
}

SFPrimitiveArrayValueList.prototype["setGridIndex"]=function(gridIndex){
	this.position=this.array.getPrimitive().getIndicesPositions()[gridIndex];
	this.arrayData=this.array.getPrimitiveData(gridIndex);
};
		
SFPrimitiveArrayValueList.prototype["generateValue"]=function(){
	return this.arrayData.generateSample();
};
		
SFPrimitiveArrayValueList.prototype["getDataIndex"]=function(index){
	return this.primitiveIndices[this.position+index];
};
		
SFPrimitiveArrayValueList.prototype["getValue"]=function(index){
	var value=this.arrayData.generateSample();
	this.arrayData.getElement(this.getDataIndex(index), value);
	return value;
};
		
